apps/src/templates/admin/MassDeleteContainer.tsx

```diff
@@ -0,0 +1,438 @@
+
+  // Keep original uploaded data in state
+  const [originalCsvData, setOriginalCsvData] = useState<OriginalData[]>();
+  // Keep processed data ready for deletion
+  const [processedCsvData, setProcessedCsvData] = useState<ProcessedData[]>();
+
+  const handleFileUpload = async (
+    event: React.ChangeEvent<HTMLInputElement>
+  ) => {
+    const file = event.target.files?.[0];
+    setCsvError(null);
+
+    if (file && file.type === 'text/csv') {
+      setCsvFile(file);
+
+      try {
+        // Parse and validate CSV
+        const parsedData = await parseCsv(file);
+        setOriginalCsvData(parsedData);
+        setCsvData(parsedData);
+
+        // Determine if conversion is needed
+        if (hasUsernameColumn(parsedData)) {
+          setHasUsernames(true);
+        } else {
+          // Convert to ProcessedData format for data that already has student_id
+          const processedData = parsedData.map(item => ({
+            student_id: item.student_id!,
+            unit_name: item.unit_name,
+          }));
+          setProcessedCsvData(processedData);
+          setHasUsernames(false);
+        }
+      } catch (error) {
+        setCsvError(
+          error instanceof Error ? error.message : 'Failed to parse CSV'
+        );
+        setCsvData(null);
+        setOriginalCsvData(undefined);
+        setProcessedCsvData(undefined);
+      }
+    } else {
+      setCsvError('Please upload a valid CSV file');
+    }
+  };
+
+  const convertUsernamesToIds = async () => {
+    if (!originalCsvData || !hasUsernames) {
+      console.log('No username data to convert');
+      return;
+    }
+
+    setIsConverting(true);
+    setConversionError(null);
+    console.log('Starting username to ID conversion...');
+
+    try {
+      const response = await fetch('/admin/convert_usernames_to_ids', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+          Accept: 'application/json',
+          'X-CSRF-Token': getCSRFToken(),
+        },
+        body: JSON.stringify({
+          csv_data: originalCsvData,
+          teacher_id: teacherId,
+        }),
+      });
+
+      if (!response.ok) {
+        console.info(
+          `HTTP error for convert_usernames_to_ids. status: ${response.status}`
+        );
+      }
+
+      const result = await response.json();
+
+      if (result.success) {
+        setProcessedCsvData(result.converted_data);
+        setCsvData(result.converted_data);
+        setHasUsernames(false);
+      }
+    } catch (error) {
+      console.error('Error during conversion:', error);
+      setConversionError(
+        error instanceof Error
+          ? error.message
+          : 'Failed to convert usernames to student IDs. Please try again.'
+      );
+    } finally {
+      setIsConverting(false);
+    }
+  };
+
+  const deleteProgress = async (isDryRun = true) => {
+    try {
+      const response = await fetch('/admin/delete_user_progress', {
```

Comment:
nit: Similar to isConverting, we may want a loading state/to disable the button so that we can't call this `deleteProgress` method multiple times before it completes.